crates/ibc/src/services/core/client.rs

```diff
@@ -0,0 +1,324 @@
+use core::str::FromStr;
+use std::boxed::Box;
+
+use ibc_proto::google::protobuf::Any;
+use ibc_proto::ibc::core::client::v1::query_server::Query as ClientQuery;
+use ibc_proto::ibc::core::client::v1::{
+    ConsensusStateWithHeight, IdentifiedClientState, Params as ClientParams,
+    QueryClientParamsRequest, QueryClientParamsResponse, QueryClientStateRequest,
+    QueryClientStateResponse, QueryClientStatesRequest, QueryClientStatesResponse,
+    QueryClientStatusRequest, QueryClientStatusResponse, QueryConsensusStateHeightsRequest,
+    QueryConsensusStateHeightsResponse, QueryConsensusStateRequest, QueryConsensusStateResponse,
+    QueryConsensusStatesRequest, QueryConsensusStatesResponse, QueryUpgradedClientStateRequest,
+    QueryUpgradedClientStateResponse, QueryUpgradedConsensusStateRequest,
+    QueryUpgradedConsensusStateResponse,
+};
+use tonic::{Request, Response, Status};
+use tracing::trace;
+
+use crate::core::ics02_client::client_state::ClientStateValidation;
+use crate::core::ics02_client::error::ClientError;
+use crate::core::ics24_host::identifier::ClientId;
+use crate::core::ics24_host::path::{
+    ClientConsensusStatePath, ClientStatePath, Path, UpgradeClientPath,
+};
+use crate::core::{ContextError, ValidationContext};
+use crate::hosts::tendermint::upgrade_proposal::UpgradeValidationContext;
+use crate::prelude::*;
+use crate::services::core::context::QueryContext;
+use crate::Height;
+
+// TODO(rano): currently the services don't support pagination, so we return all the results.
+
+pub struct ClientQueryService<I, U>
+where
+    I: QueryContext + Send + Sync + 'static,
+    U: UpgradeValidationContext + Send + Sync + 'static,
+    <I as ValidationContext>::AnyClientState: Into<Any>,
+    <I as ValidationContext>::AnyConsensusState: Into<Any>,
+    <U as UpgradeValidationContext>::AnyClientState: Into<Any>,
+    <U as UpgradeValidationContext>::AnyConsensusState: Into<Any>,
+{
+    ibc_context: I,
```

Comment:
I think we want to use an `Arc<Mutex<I>>` here somehow (might require changing a few other things). The general idea is that tonic spawns some threads that will be reading, and the main thread with the `ExecutionContext` will be writing, so they need to be coordinated with a mutex.

Gaining ownership of the `ValidationContext` doesn't work, or at least is at lot more work for the user to get right (see for example [this](https://github.com/informalsystems/basecoin-rs/pull/125/files#r1321918507) basecoin-rs comment).

Reply:
I put `Send + Sync + 'static` bound on `I` already. Doesn't it already guarantee the same behavior of `Arc<Mutex<I>>`?

Reply:
Regarding the `.clone()` on basecoin-rs side, in my opinion, it's an implementation bug in my PR. Let me think a little bit more. @Farhad-Shabani told me, it may make sense to rather make `IbcContext` to satisfy `Send + Sync + 'static`.

Reply:
For additional context, within the `basecoin-rs`, anything wrapped with structs like `SharedStore` that implements **[Store](https://github.com/informalsystems/basecoin-rs/blob/04b549e9d9408a040d5ef665a2019d77d0539fae/crates/store/src/context.rs#L8)** type already adheres to the `Send + Sync + 'static` requirements. Though, exactly... there are certain fields, such as **[`client_counter`](https://github.com/informalsystems/basecoin-rs/blob/e9df0f01b174fa589f6c177fd53d2b7a131d6bf3/crates/app/src/modules/ibc/impls.rs#L295-L296)**, that do not possess this property. I think this can be resolved by wrapping them as `Arc<Mutex<u64>>`.
This issue also became particularly challenging for me when I was implementing the mock Cosmos chain for the relayer-next, and there were places I needed to safely clone the `ibc_context`.

Reply:
> Though, exactly... there are certain fields, such as [client_counter](https://github.com/informalsystems/basecoin-rs/blob/e9df0f01b174fa589f6c177fd53d2b7a131d6bf3/crates/app/src/modules/ibc/impls.rs#L295-L296), that do not possess this property.

`client_counter` is a `u64`, which is `Send + Sync + 'static`.

> I put `Send + Sync + 'static` bound on `I` already. Doesn't it already guarantee the same behavior of `Arc<Mutex<I>>`?

No, all it ensures it that `I` (and `&I`) can be sent across threads. Any type that can be sent across threads can be used there, not just `Arc<Mutex<_>>`. For example, `String`, `HashMap<String, u64`, `u64`, etc.

> Regarding the .clone() on basecoin-rs side, in my opinion, it's an implementation bug in my PR.

Yes it is, but your API in ibc-rs allowed for that bug. Now either we leave it as is, but heavily document how to properly use it; it must be `Arc<Mutex<_>>`, `Arc<RwLock<_>>`, or any type with equivalent behavior. The important property is that writes done in the main thread must be seen by other threads, and `Send + Sync + 'static` doesn't ensure that (as we see with your basecoin PR). Or, we ensure that property is held by forcing the user to give us a `Arc<Mutex<_>>` or `Arc<RwLock<_>>`, which is the direction which I originally was suggesting.

Reply:
> client_counter is a u64, which is Send + Sync + 'static.

> The important property is that writes done in the main thread must be seen by other threads

Ah, yes, my mistake! The above is the needed property, so I was suggesting to use the `Arc<Mutex<u64>` for the `client_coutner`

Reply:
There are [`Query::from_arc()`](https://docs.rs/ibc-proto/0.34.1/ibc_proto/ibc/core/client/v1/msg_server/struct.MsgServer.html#method.from_arc) for these service traits. Can we somehow leverage that?

Reply:
We can probably leave as is, but add a big docstring for all types that take a `ibc_context: I` about how `I` should be `Arc<Mutex>` or whatever allows reads to see writes (as discussed in [here](https://github.com/cosmos/ibc-rs/pull/842#discussion_r1323362097)).

Reply:
I added doc strings accordingly in my last commit. Let me know if they are okay.